### PR TITLE
Add  setting to ignore throttling nodes for allocation of unassigned …

### PR DESCRIPTION
**Pull request by gbbafna**
…primaries in restore in order to speed up remote restore

### Description

For remote store backed domains, if a node in the cluster is replaced by a new node and we trigger restore for all the red indices,  all the recoveries are happening only on the new node . 

This allocation to only one node is due to balancer weights as well multiple Allocation Constraints, which provides weight to each node. We will assign shards only to the node with minimum weight . The new node will always have minimum weight for all the unassigned shards.  The new node will keep throttling the recoveries, until the existing recoveries are done.  This results in slowing down the overall recovery time. 

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [x] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java [hunk 1]
```diff
@@ -154,6 +154,13 @@ public class BalancedShardsAllocator implements ShardsAllocator {
         Property.NodeScope
     );
 
+    public static final Setting<Boolean> IGNORE_THROTTLE_FOR_REMOTE_RESTORE = Setting.boolSetting(
+        "cluster.routing.allocation.ignore_throttle_for_restore",
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:158]: Let's rename this better to `cluster.routing.allocation.recovery.remote_primary.ignore_balance `

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java [hunk 2]
```diff
@@ -918,7 +923,15 @@ AllocateUnassignedDecision decideAllocateUnassigned(final ShardRouting shard) {
                 nodeExplanationMap.put(node.getNodeId(), new NodeAllocationResult(node.getRoutingNode().node(), currentDecision, 0));
                 nodeWeights.add(Tuple.tuple(node.getNodeId(), currentWeight));
             }
-            if (currentDecision.type() == Decision.Type.YES || currentDecision.type() == Decision.Type.THROTTLE) {
+
+            // For REMOTE_STORE recoveries, THROTTLE is as good as NO as we want faster recoveries
+            // The side effect of this are increased relocations post these allocations.
+            boolean considerThrottleAsNo = ignoreThrottleInRestore
+                && shard.recoverySource().getType() == RecoverySource.Type.REMOTE_STORE
+                && shard.primary();
```
Comment (vikasvb90) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java:931]: Do you think there should still be a cap on the maximum restores? Say on the basis of in-flight shard sizes ?

Comment (gbbafna) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java:931]: There is a cap on maximum restores dictated by `max_concurrent_recoveries` . This change just ignore the nodes which are "THROTTLING` .

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java [hunk 3]
```diff
@@ -918,7 +923,15 @@ AllocateUnassignedDecision decideAllocateUnassigned(final ShardRouting shard) {
                 nodeExplanationMap.put(node.getNodeId(), new NodeAllocationResult(node.getRoutingNode().node(), currentDecision, 0));
                 nodeWeights.add(Tuple.tuple(node.getNodeId(), currentWeight));
             }
-            if (currentDecision.type() == Decision.Type.YES || currentDecision.type() == Decision.Type.THROTTLE) {
+
+            // For REMOTE_STORE recoveries, THROTTLE is as good as NO as we want faster recoveries
+            // The side effect of this are increased relocations post these allocations.
+            boolean considerThrottleAsNo = ignoreThrottleInRestore
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java:929]: Could this change regardless of remote store or not?

**Review by ashking94 — COMMENTED** [review]
lgtm

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java (hunk 3 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java:929]: I would rather start with remote store and later extend to snapshot restore if needed .

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java (hunk 2 above)
Comment (vikasvb90) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java:931]: Right, makes sense!

File: server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+            }
+        };
+        List<AllocationDecider> allocationDeciders = Arrays.asList(rebalanceDecider, allocationDecider);
+        RoutingAllocation routingAllocation = newRoutingAllocation(new AllocationDeciders(allocationDeciders), clusterState);
+        // allocate and get the node that is now relocating
+        Settings build = Settings.builder().put(IGNORE_THROTTLE_FOR_REMOTE_RESTORE.getKey(), true).build();
+        BalancedShardsAllocator allocator = new BalancedShardsAllocator(build);
+        allocator.allocate(routingAllocation);
+        assertEquals(routingAllocation.routingNodes().shardsWithState(ShardRoutingState.INITIALIZING).get(0).currentNodeId(), "node_0");
+        ;
```
Comment (ashking94) [inline server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java:82]: pls remove this unnecessary line.

File: server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+        AllocationDecider allocationDecider = new AllocationDecider() {
+            @Override
+            public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+                if (throttleNodes.contains(node.nodeId())) {
+                    return Decision.THROTTLE;
+                }
+                return Decision.YES;
+            }
+        };
+        AllocationDecider rebalanceDecider = new AllocationDecider() {
```
Comment (ashking94) [inline server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java:69]: Can you explain why do we need to add a rebalanceDecider here?

File: server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+                return Decision.YES;
+            }
+        };
+        List<AllocationDecider> allocationDeciders = Arrays.asList(rebalanceDecider, allocationDecider);
+        RoutingAllocation routingAllocation = newRoutingAllocation(new AllocationDeciders(allocationDeciders), clusterState);
+        // allocate and get the node that is now relocating
+        Settings build = Settings.builder().put(IGNORE_THROTTLE_FOR_REMOTE_RESTORE.getKey(), true).build();
+        BalancedShardsAllocator allocator = new BalancedShardsAllocator(build);
+        allocator.allocate(routingAllocation);
+        assertEquals(routingAllocation.routingNodes().shardsWithState(ShardRoutingState.INITIALIZING).get(0).currentNodeId(), "node_0");
```
Comment (ashking94) [inline server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java:81]: Should we also assert that the initialising shard is of `idx2`?

File: server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+import java.util.stream.Collectors;
+
+import static org.opensearch.cluster.metadata.IndexMetadata.SETTING_CREATION_DATE;
+import static org.opensearch.cluster.metadata.IndexMetadata.SETTING_NUMBER_OF_REPLICAS;
+import static org.opensearch.cluster.metadata.IndexMetadata.SETTING_NUMBER_OF_SHARDS;
+import static org.opensearch.cluster.metadata.IndexMetadata.SETTING_VERSION_CREATED;
+import static org.opensearch.cluster.routing.allocation.allocator.BalancedShardsAllocator.IGNORE_THROTTLE_FOR_REMOTE_RESTORE;
+
+public class DecideAllocateUnassignedTests extends OpenSearchAllocationTestCase {
+    public void testSingleShardBalanceProducesSameResultsAsBalanceStep_IgnoreThrottle() {
```
Comment (ashking94) [inline server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java:48]: nit - could we rename the test to better reflect the purpose of the test? I was slighly confused by reading the name and seeing the test.

File: server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+            public Decision canRebalance(ShardRouting shardRouting, RoutingAllocation allocation) {
+                return Decision.YES;
+            }
+        };
+        List<AllocationDecider> allocationDeciders = Arrays.asList(rebalanceDecider, allocationDecider);
+        RoutingAllocation routingAllocation = newRoutingAllocation(new AllocationDeciders(allocationDeciders), clusterState);
+        // allocate and get the node that is now relocating
+        Settings build = Settings.builder().put(IGNORE_THROTTLE_FOR_REMOTE_RESTORE.getKey(), false).build();
+        BalancedShardsAllocator allocator = new BalancedShardsAllocator(build);
+        allocator.allocate(routingAllocation);
```
Comment (ashking94) [inline server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java:118]: Much of the code between the 2 tests are duplicate. Can we extract a method and reuse the same?

File: server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+                    .put(SETTING_NUMBER_OF_SHARDS, 1)
+                    .put(SETTING_NUMBER_OF_REPLICAS, 0)
+                    .put(SETTING_CREATION_DATE, System.currentTimeMillis())
+            )
+            .primaryTerm(0, primaryTerm)
+            .build();
+
+        IndexShardRoutingTable.Builder indexShardRoutingBuilder = new IndexShardRoutingTable.Builder(shardId);
+        String primaryNode = null;
+        String relocatingNode = null;
```
Comment (ashking94) [inline server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java:148]: Why do we need to declare this when we are anyways passing null values to the method hereon?

File: server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+    }
+
+    private ClusterState addNodesToClusterState(ClusterState clusterState, int nodeId) {
+        DiscoveryNodes.Builder nodesBuilder = DiscoveryNodes.builder(clusterState.nodes());
+        DiscoveryNode discoveryNode = newNode("node_" + nodeId);
+        nodesBuilder.add(discoveryNode);
+        return ClusterState.builder(clusterState).nodes(nodesBuilder).build();
+    }
+
+    private ClusterState addRestoringIndexToClusterState(ClusterState clusterState, String index) {
```
Comment (ashking94) [inline server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java:131]: Are there any existing methods elsewhere that can be reused here as well?

**Review by ashking94 — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java (hunk 1 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java]: Renamed to `cluster.routing.allocation.remote_primary.ignore_throttle`

File: server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java (hunk 5 above)
Comment (gbbafna) [reply server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java]: Removed it . Not needed

File: server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java (hunk 6 above)
Comment (gbbafna) [reply server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java:64]: Done

File: server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java (hunk 7 above)
Comment (gbbafna) [reply server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java]: Done.

File: server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java (hunk 8 above)
Comment (gbbafna) [reply server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java]: Extracted allocation decider code.

File: server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java (hunk 9 above)
Comment (gbbafna) [reply server/src/test/java/org/opensearch/cluster/routing/allocation/DecideAllocateUnassignedTests.java]: Fixed now

**Review by ashking94 — APPROVED** [review]
LGTM

**Pull request merged**